SourceX/controls/plrctrls.cpp

```diff
@@ -0,0 +1,689 @@
+#include "controls/plrctrls.h"
+
+#include <cstdint>
+
+#include "controls/controller_motion.h"
+#include "controls/game_controls.h"
+
+// Based on the Nintendo Switch port by @lantus, @erfg12, @rsn8887.
+
+namespace dvl {
+
+#define INV_TOP 240;
+#define INV_LEFT RIGHT_PANEL_X;
+#define INV_HEIGHT 320;
+
+bool newCurHidden = false;
+coords speedspellscoords[50];
+int speedspellcount = 0;
+
+// Native game menu, controlled by simulating a keyboard.
+bool InGameMenu()
+{
+	return stextflag > 0 || questlog || helpflag || talkflag || qtextflag || sgpCurrentMenu;
+}
+
+namespace {
+
+DWORD invmove = 0;
+int hsr[3] = { 0, 0, 0 }; // hot spell row counts
+int slot = SLOTXY_INV_FIRST;
+int spbslot = 0;
+
+// Menu controlled by simulating a mouse.
+bool InControlledMenu()
+{
+	return invflag || spselflag || chrflag;
+}
+
+// 0 = not near, >0 = distance related player 1 coordinates
+coords checkNearbyObjs(int x, int y, int diff)

```

Comment:
This could be improved https://stackoverflow.com/questions/30368632/calculate-distance-on-a-grid-between-2-points

But even with that it's not taking walls, objects or direction of player in to account. So maybe it would be better to use the path finding algorithm. Or even a combination (first check if it is close, before trying to calculate a path) for performance.